src/realm/object-store/sync/app.cpp

```diff
@@ -862,185 +908,221 @@ std::string App::get_app_route(const Optional<std::string>& hostname) const
         }
         else {
-            route = util::format("%1/location", new_hostname ? get_app_route(new_hostname) : get_app_route());
+            base_url = new_hostname ? *new_hostname : m_base_url;
+            // If this is for a redirect after querying new_hostname, then use the redirect location
+            if (redir_location)
+                app_route = get_app_route(redir_location);
+            // If this is querying the new_hostname, then use that location
+            else if (new_hostname)
+                app_route = get_app_route(new_hostname);
+            else
+                app_route = get_app_route();
+            REALM_ASSERT(!app_route.empty());
         }
     }
+
     Request req;
     req.method = HttpMethod::get;
-    req.url = route;
+    req.url = util::format("%1/location", app_route);
     req.timeout_ms = m_request_timeout_ms;
+    req.redirect_count = redirect_count;
 
-    log_debug("App: request location: %1", route);
+    log_debug("App: request location: %1", req.url);
 
-    m_config.transport->send_request_to_server(req, [self = shared_from_this(),
-                                                     completion = std::move(completion)](const Response& response) {
-        // If the response contains an error, then pass it up
-        if (response.http_status_code >= 300 || (response.http_status_code < 200 && response.http_status_code != 0)) {
-            return completion(response); // early return
-        }
-
-        try {
-            auto json = parse<BsonDocument>(response.body);
-            auto hostname = get<std::string>(json, "hostname");
-            auto ws_hostname = get<std::string>(json, "ws_hostname");
-            auto deployment_model = get<std::string>(json, "deployment_model");
-            auto location = get<std::string>(json, "location");
-            if (self->m_sync_manager->perform_metadata_update([&](SyncMetadataManager& manager) {
-                    manager.set_app_metadata(deployment_model, location, hostname, ws_hostname);
-                })) {
-                // Update the hostname and sync route using the new app metadata info
-                self->update_hostname(self->m_sync_manager->app_metadata());
-            }
-            else {
-                // No metadata in use, update the hostname and sync route directly
-                self->update_hostname(hostname, ws_hostname);
+    m_config.transport->send_request_to_server(
+        std::move(req), [self = shared_from_this(), completion = std::move(completion),
+                         base_url = std::move(base_url)](Request&& request, const Response& response) mutable {
+            // Check to see if a redirect occurred
+            if (AppUtils::is_redirect_status_code(response.http_status_code)) {
+                // Make sure we don't do too many redirects (max_http_redirects (20) is an arbitrary number)
+                if (++request.redirect_count >= max_http_redirects) {
```

Comment:
shouldn't this check be strictly greater than?

Reply:
Actually, the redirect count matches the number of redirects that occurred, so when `request.redirect_count == 20`, then 20 successive redirects have already happened.